system/security/plugins/jwtSecurity/jwtCache.hpp

```diff
@@ -0,0 +1,455 @@
+
+        /**
+         * Erase a file scope value from the object.  Does nothing when the given scope
+         * does not exist.
+         *
+         * @param   scope           A file scope pattern
+         *
+         * @return  Reference to updated object (this).
+         *
+         * @see     clearFileScopePerm
+         */
+        JWTUserInfo& eraseFileScopePerm(const std::string& scope);
+
+        /**
+         * Erase all file scope values from the object.
+         *
+         * @return  Reference to updated object (this).
+         *
+         * @see     eraseFileScopePerm
+         */
+        JWTUserInfo& clearFileScopePerm();
+
+        /**
+         * @return  The number of file scopes currently saved.
+         */
+        unsigned int countFileScopePerms() const;
+
+        /**
+         * @return  True if any file scopes have been set, false otherwise.
+         */
+        bool hasFileScopePerms() const;
+
+    private:
+
+        std::string _globToExample(const std::string& pattern) const;
+        std::string _convertPathname(const std::string& logicalPath) const;
+        JWTUserInfo& _addScopePerm(PermissionMap& scopePermMap, const std::string& scope, SecAccessFlags accessFlag, SecAccessFlags defaultFlag, bool asDeny);
+        JWTUserInfo& _addScopePerm(PermissionMap& scopePermMap, const std::string& scope, const std::string& accessFlagName, SecAccessFlags defaultFlag, bool asDeny);
+        SecAccessFlags _matchScopePerm(const PermissionMap& scopePermMap, const std::string& scope, SecAccessFlags defaultFlag) const;
+        JWTUserInfo& _eraseScopePerm(PermissionMap& scopePermMap, const std::string& scope);
+        JWTUserInfo& _clearScopePerm(PermissionMap& scopePermMap);
+        unsigned int _countScopePerms(const PermissionMap& scopePermMap) const;
+
+    private:
+
+        time_t                  expireTimeUTC;      //!< Time this object expires
+        std::string             refreshToken;       //!< The refresh token associated with jwtToken
+        std::string             jwtToken;           //!< Raw token provided during authentication; suitable for passing to another service
+        PermissionMap           featurePermMap;     //!< Map of FeatureName -> PermissionFlag
+        PermissionMap           wuScopePermMap;     //!< Map of ScopePattern -> PermissionFlag
+        PermissionMap           fileScopePermMap;   //!< Map of ScopePattern -> PermissionFlag
+};
+
+/**
+ * This is basically a thread-safe map of username->JWTUserInfo objects.
+ */
+class JWTUserCache
+{
+    private:
+
+        typedef std::map<std::string, std::shared_ptr<JWTUserInfo> > UserPermissionMap;
+
+    public:
+
+        bool has(const std::string& userName) const;
+
+        JWTUserCache& set(const std::string& userName, std::shared_ptr<JWTUserInfo>& userInfo);
+
+        const std::shared_ptr<JWTUserInfo> get(const std::string& userName);
```

Comment:
I should have been clearer.  I meant 

`    std::shared_ptr<JWTUserInfo> get(const std::string& userName) const;
`

Reply:
Done, and changed to using a const_iterator as well.